### PR TITLE
Improve termination statuses

### PR DESCRIPTION
**Pull request by guimarqu**
I kept `NODE_LIMIT` because it can be usefull when we call subsolvers through MOI, or when we only solve the root node.

I replaced `NOT_YET_DETERMINED`  by  `UNKOWN_TERMINATION` (symmetric to `UNKNOWN_FEASIBILITY`) & `UNDEFINED`.

### CODE REVIEW

**Comment by rrsadykov** [comment]
Why don't we use the termination status only? It can return also infeasible. 

I have checked : there is a single termination status in MOI, which may return INFEASIBLE. May be we just use `TerminationStatusCode` of MOI?

**Comment by guimarqu** [comment]
> Why don't we use the termination status only? It can return also infeasible.
> 
> I have checked : there is a single termination status in MOI, which may return INFEASIBLE. May be we just use `TerminationStatusCode` of MOI?

We don't use MOI TerminationStatusCode because otherwise a package that contains an algorithm for Coluna will have MOI in its dependencies.

We separated Termination & Feasibility because Feasiblity can be also used for a solution (not only for the problem).

**Comment by rrsadykov** [comment]
> We don't use MOI TerminationStatusCode because otherwise a package that contains an algorithm for Coluna will have MOI in its dependencies.

If a package depends on Coluna, it automatically depends on MOI, no? 
 
> We separated Termination & Feasibility because Feasibility can be also used for a solution (not only for the problem).

Nothing prevents us from using different statuses for algorithms and solutions. In MOI, they use different statuses. 
It is misleading for me to have two statuses. If feasibility status is infeasible, which value should the termination status take?

**Comment by guimarqu** [comment]
> If a package depends on Coluna, it automatically depends on MOI, no?

Yes but this dependence won't be in the Project file but only in the Manifest. It's a "weaker" dependence.
 
> Nothing prevents us from using different statuses for algorithms and solutions. In MOI, they use different statuses.
> It is misleading for me to have two statuses for the algorithm. If feasibility status is infeasible, which value should the termination status take?

Yes I agree, we go for a `TerminationStatus` and `SolutionStatus` ?

`TerminationStatus` : `OPTIMAL`, `INFEASIBLE`, `TIME_LIMIT`, `NODE_LIMIT`, `OTHER_LIMIT`, `UNKNOWN_TERMINATION_STATUS`, `UNDEFINED_TERMINATION_STATUS`

`SolutionStatus` : `FEASIBLE_SOL`, `INFEASIBLE_SOL`, `EMPTY_SOLUTION`,  `UNKNOWN_SOLUTION_STATUS`, `UNDEFINED_SOLUTION_STATUS`

**Comment by rrsadykov** [comment]
Ok for me, just one question: what is the difference between UNKNOWN_TERMINATION_STATUS and UNDEFINED_TERMINATION_STATUS?

**Comment by guimarqu** [comment]
> Ok for me, just one question: what is the difference between UNKNOWN_TERMINATION_STATUS and UNDEFINED_TERMINATION_STATUS?

`UNKOWN` it's before the algorithm, `UNDEFINED` it's after the algorithm & it's used only when the subsolver is called through MOI and that MOI returns a status that we do not cover. Maybe we can call it `UNCOVERED`...

**Comment by guimarqu** [comment]
I'll try to avoid the use of UNKNOWN. If we don't need it, I'll remove it.

**Comment by rrsadykov** [comment]
Ok for me

**Comment by rrsadykov** [comment]
I did not understand, why you set the solution infeasible status in the Benders algorithm, and termination infeasible status in for the column generation (in the case of infeasibility)?

**Comment by guimarqu** [comment]
I have a doubt about statuses in Benders, I'll check & finish tomorrow.
I think it's ok for column generation.

Guillaume Marques

**Comment by rrsadykov** [comment]
Ok. I also think that you should set the solution status when you change the solution in `OptimizationState`.

**Comment by guimarqu** [comment]
> Ok. I also think that you should set the solution status when you change the solution in `OptimizationState`.

How do you know if the solution pushed into the `OptimizationState` is feasible or not ?

**Comment by rrsadykov** [comment]
> > Ok. I also think that you should set the solution status when you change the solution in `OptimizationState`.
> 
> How do you know if the solution pushed into the `OptimizationState` is feasible or not ?

The algorithm which creates the solution should know this.

**Comment by rrsadykov** [comment]
Also, there are many solutions in `OptimizationState`, the solution status refers to which one?

**Comment by guimarqu** [comment]
Ok so I suggest to move the `SolutionStatus` into `Solution`.
It will mean that the solution is (in)feasible with regard to the model referenced.

Then, we'll have to give the status of the solution at its instantiation.

File: src/Algorithm/utilities/optimizationstate.jl [hunk 1]
```diff
@@ -335,7 +336,7 @@ end
 function Base.print(io::IO, form::AbstractFormulation, optstate::OptimizationState)
     println(io, "┌ Optimization state ")
     println(io, "│ Termination status: ", optstate.termination_status)
-    println(io, "│ Feasibility status: ", optstate.feasibility_status)
+    println(io, "│ Solution status: ", optstate.solution_status)
```
Comment (rrsadykov) [inline src/Algorithm/utilities/optimizationstate.jl:339]: No more solution status for `OptimizationState`, no?

**Comment by rrsadykov** [comment]
Otherwise, this merge is ok for me.

File: src/Algorithm/utilities/optimizationstate.jl (hunk 1 above)
Comment (guimarqu) [reply src/Algorithm/utilities/optimizationstate.jl:339]: I'll fix it.

File: src/Algorithm/colgenstabilization.jl [hunk 2]
```diff
@@ -105,7 +105,7 @@ function linear_combination(in_dual_sol::DualSolution, out_dual_sol::DualSolutio
     for (i, constrid) in enumerate(constrids)
         bound += constrvals[i] * getcurrhs(form, constrid) 
     end
-    return DualSolution(form, constrids, constrvals, bound)
+    return DualSolution(form, constrids, constrvals, bound, FEASIBLE_SOL)
```
Comment (guimarqu) [inline src/Algorithm/colgenstabilization.jl:108]: @rrsadykov I'm not sure about this `SolutionStatus`. Does this solution require a status ?

File: src/Algorithm/colgenstabilization.jl [hunk 3]
```diff
@@ -144,7 +144,7 @@ function update_alpha_automatically!(
 
     # first we calculate the in-sep direction
     constrids, constrvals = componentwisefunction(smooth_dual_sol, storage.stabcenter, -)
-    in_sep_direction = DualSolution(master, constrids, constrvals, 0.0)
+    in_sep_direction = DualSolution(master, constrids, constrvals, 0.0, FEASIBLE_SOL)
```
Comment (guimarqu) [inline src/Algorithm/colgenstabilization.jl:147]: Same question here.

File: src/Algorithm/colgenstabilization.jl [hunk 4]
```diff
@@ -158,7 +158,7 @@ function update_alpha_automatically!(
             push!(constrrhs, getcurrhs(master, constrid))
         end 
     end
-    subgradient = DualSolution(master, constrids, constrrhs, 0.0)
+    subgradient = DualSolution(master, constrids, constrrhs, 0.0, FEASIBLE_SOL)
```
Comment (guimarqu) [inline src/Algorithm/colgenstabilization.jl:161]: Same here.

File: src/Algorithm/colgenstabilization.jl (hunk 2 above)
Comment (rrsadykov) [reply src/Algorithm/colgenstabilization.jl:108]: You can put the solutions status to something like "unknown_feasibility" for the dual solutions used for stabilization.

**Comment by guimarqu** [comment]
Ready to merge, waiting for approvation.

**Comment by rrsadykov** [comment]
Ok for me. However, it seems I do not have rights to approve this merge.

**Comment by guimarqu** [comment]
> Ok for me. However, it seems I do not have rights to approve this merge.

But you are in the list of possible reviewers.

**Review by rrsadykov — APPROVED** [review]

**Comment by rrsadykov** [comment]
Ok, I found.

**Pull request merged**